components/tools/annotation.js

```diff
@@ -3,95 +3,53 @@
-            if (!arguments.length) {
-                return yScale;
-            }
-            yScale = value;
-            return annotation;
-        };
-
-        annotation.yValue = function(value) {
-            if (!arguments.length) {
-                return yValue;
-            }
-            yValue = value;
-            return annotation;
-        };
-
-        annotation.yLabel = function(value) {
-            if (!arguments.length) {
-                return yLabel;
-            }
-            yLabel = value;
-            return annotation;
-        };
-
-        annotation.padding = function(value) {
-            if (!arguments.length) {
-                return padding;
-            }
-            padding = value;
-            return annotation;
-        };
-
-        annotation.formatCallout = function(value) {
-            if (!arguments.length) {
-                return formatCallout;
-            }
-            formatCallout = value;
-            return annotation;
-        };
+            selection.each(function(data) {
+                var x = annotation.xScale.value,
+                    y = annotation.yScale.value,
+                    yValue = annotation.yValue.value;
+
+                var container = d3.select(this);
+
+                // Create a group for each annotation
+                var g = fc.utilities.simpleDataJoin(container, 'annotation', data, yValue);
+
+                // Added the required elements - each annotation consists of a line and text label
+                var enter = g.enter();
+                enter.append('line')
+                    .attr('class', 'marker');
+                enter.append('text')
+                    .attr('class', 'label');
+
+                // Update the line
+                g.selectAll('line.marker').attr('x1', x.range()[0])
+                    .attr('y1', function(d) {
+                        return y(yValue(d));
+                    })
+                    .attr('x2', x.range()[1])
+                    .attr('y2', function(d) {
+                        return y(yValue(d));
+                    });
+
+                // Update the text label
+                var paddingValue = annotation.padding.value.apply(this, arguments);
+                g.selectAll('text.label').attr('x', x.range()[1] - paddingValue)
+                    .attr('y', function(d) {
+                        return y(yValue(d)) - paddingValue;
+                    })
+                    .attr('style', 'text-anchor: end;')
+                    .text(annotation.formatLabel.value);
+
+                annotation.decorate.value(container);
+            });
+        };
+
+        annotation.xScale = fc.utilities.property(d3.time.scale());
+        annotation.yScale = fc.utilities.property(d3.scale.linear());
+        annotation.yValue = fc.utilities.functorProperty(fc.utilities.fn.identity);
+        annotation.formatLabel = fc.utilities.functorProperty(annotation.yValue.value);
```

Comment:
Ops, memory fail! Agreed label is better though.

On Mon, 19 Jan 2015 12:25 James Leftley <notifications@github.com> wrote:

> In components/tools/annotation.js
> <https://github.com/ScottLogic/d3-financial-components/pull/256#discussion-diff-23157087>
> :
>
> > +                var paddingValue = annotation.padding.value.apply(this, arguments);
> > +                g.selectAll('text.label').
>
> attr('x', x.range()[1] - paddingValue)
> > +                    .attr('y', function(d) {
> > +                        return y(yValue(d)) - paddingValue;
> > +                    })
> > +                    .attr('style', 'text-anchor: end;')
> > +                    .text(annotation.formatLabel.value);
> > +
> > +                annotation.decorate.value(container);
> > +            });
> > +        };
> > +
> > +        annotation.xScale = fc.utilities.property(d3.time.scale());
> > +        annotation.yScale = fc.utilities.property(d3.scale.linear());
> > +        annotation.yValue = fc.utilities.functorProperty(fc.utilities.fn.identity);
> > +        annotation.formatLabel = fc.utilities.functorProperty(annotation.yValue.value);
>
> This was based on #225
> <https://github.com/ScottLogic/d3-financial-components/issues/225>
>
> I think label would be more appropriate.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ScottLogic/d3-financial-components/pull/256/files#r23157087>
> .
>